### PR TITLE
Add subject alternative names to azurerm_key_vault_certificate

### LINKED ISSUE #479: `azurerm_key_vault_certificate` - support for speciying the `subject_alternative_name`
Currently the "azurerm_key_vault_certificate" resource does not allow for the creation of a SubjectAlternativeName within the X509CertificateProperties object. It would be great if the resource could be extended to allow for this.

### LINKED ISSUE #1309: Specifying Subject Alternative Name in Azure KeyVault Certificate resource
Hi,

Is there any way I can specify Specifying Subject Alternative Name in Azure Keyvault resource for certificate generation.

### PR DESCRIPTION
**Pull request by draggeta**
Because of hacktober I decided to just try my hand at this. Either way. This allows you to specify subject alternative names for the azurerm_key_vault_certificate resource. All combinations of FQDNs, email addresses and UPNs are possible.

Test results:
```
PS > go test -timeout 180m -v -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc" github.com\terraform-providers\terraform-provider-azurerm\azurerm -run ^TestAccAzureRMKeyVaultCertificate_basicGenerateSans$
=== RUN   TestAccAzureRMKeyVaultCertificate_basicGenerateSans
--- PASS: TestAccAzureRMKeyVaultCertificate_basicGenerateSans (222.49s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm       232.421s
```

Closes #479
Also fixes #1309

### CODE REVIEW

**Comment by draggeta** [comment]
This is now somehow failing on linting with `gofmt -s`. I've written to the file with `-w` and it still outputs no diff in git for me. Both on a Ubuntu and Windows device.

File: azurerm/resource_arm_key_vault_certificate.go [hunk 1]
```diff
@@ -616,15 +616,15 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
 		sanOutput := make(map[string]interface{}, 0)
 		if san := props.SubjectAlternativeNames; san != nil {
 			if emails := san.Emails; emails != nil {
-				sanOutput["emails"] = []string(*san.Emails)
+				sanOutput["emails"] = *san.Emails
```
Comment (katbyte) [inline azurerm/resource_arm_key_vault_certificate.go]: this could be `*emails`

File: azurerm/resource_arm_key_vault_certificate.go [hunk 2]
```diff
@@ -616,15 +616,15 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
...
 			}
 
 			if dnsNames := san.DNSNames; dnsNames != nil {
-				sanOutput["dns_names"] = []string(*san.DNSNames)
+				sanOutput["dns_names"] = *san.DNSNames
```
Comment (katbyte) [inline azurerm/resource_arm_key_vault_certificate.go]: this could be `*dnsNames`

File: azurerm/resource_arm_key_vault_certificate.go [hunk 3]
```diff
@@ -616,15 +616,15 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
...
 			}
 
 			if upns := san.Upns; upns != nil {
-				sanOutput["upns"] = []string(*san.Upns)
+				sanOutput["upns"] = *san.Upns
```
Comment (katbyte) [inline azurerm/resource_arm_key_vault_certificate.go]: this could be `*upns`

File: azurerm/resource_arm_key_vault_certificate.go [hunk 4]
```diff
@@ -474,17 +508,49 @@ func expandKeyVaultCertificatePolicy(d *schema.ResourceData) keyvault.Certificat
...
 			KeyUsage:         &keyUsage,
 			Ekus:             extendedKeyUsage,
+			SubjectAlternativeNames: subjectAlternativeNames,
 		}
 	}
 
 	return policy
 }
 
+func expandKeyVaultSanProperty(input []interface{}) *[]string {
```
Comment (katbyte) [inline azurerm/resource_arm_key_vault_certificate.go]: There is a `ultils.ExpandStringArray` that does exactly this

**Comment by katbyte** [comment]
Hi @draggeta,

We have a `make fmt` command we use to run, an gofmt and a `make lint` CI runs to check `gofmt` and more

**Comment by tombuildsstuff** [comment]
@draggeta we're currently using Go 1.10 on this project (but Go 1.11 is the latest) - we'll be moving to this in the near future; of particular note between the two versions is how `go fmt` can return different results (which I believe may be causing the issue here?)

File: azurerm/resource_arm_key_vault_certificate.go [hunk 5]
```diff
@@ -549,13 +615,31 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
 		for _, usage := range *props.KeyUsage {
 			usages = append(usages, string(usage))
 		}
+
+		sanOutput := make(map[string]interface{}, 0)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_key_vault_certificate.go]: can we add a list here to allow us to set this to an empty list if it's not set

```suggestion
		sanOutputs := make([]interface{}, 0)
```

File: azurerm/resource_arm_key_vault_certificate.go [hunk 6]
```diff
@@ -549,13 +615,31 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
...
+		if san := props.SubjectAlternativeNames; san != nil {
+			if emails := san.Emails; emails != nil {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_key_vault_certificate.go]: which would mean we'd add the `map[string]interface{}` inside the if statement:

```suggestion
			sanOutput := make(map[string]interface{}, 0)
			if emails := san.Emails; emails != nil {
```

File: azurerm/resource_arm_key_vault_certificate.go [hunk 7]
```diff
@@ -549,13 +615,31 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
...
+				sanOutput["emails"] = *emails
+			}
+
+			if dnsNames := san.DNSNames; dnsNames != nil {
+				sanOutput["dns_names"] = *dnsNames
+			}
+
+			if upns := san.Upns; upns != nil {
+				sanOutput["upns"] = *upns
+			}
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_key_vault_certificate.go]: and then append to it at the end:

```suggestion
			}
			
			sanOutputs = append(sanOutputs, sanOutput)
```

File: azurerm/resource_arm_key_vault_certificate.go [hunk 8]
```diff
@@ -549,13 +615,31 @@ func flattenKeyVaultCertificatePolicy(input *keyvault.CertificatePolicy) []inter
...
+		}
+
 		certProps["key_usage"] = usages
 		certProps["subject"] = *props.Subject
 		certProps["validity_in_months"] = int(*props.ValidityInMonths)
 		if props.Ekus != nil {
 			certProps["extended_key_usage"] = props.Ekus
 		}
-
+		if len(sanOutput) > 0 {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_key_vault_certificate.go]: these changes above allow us to ensure that `certProps["subject_alternative_names"]` is set either case (either to an empty list or to a list containing the item (as here). as such (with the change above) we should be able to make this:

```suggestion
					certProps["subject_alternative_names"] = sanOutputs
```

**Comment by draggeta** [comment]
@tombuildsstuff Well, that explains a lot, I'm using 1.11 😄 

As for the changes. I implemented them, but now I always end up with 1 item for the SAN when importing. Bear with me while I try and fix that.

**Comment by tombuildsstuff** [comment]
@draggeta heh :)

> Am I correct in my assumption that this should be fixed with a DiffSuppressFunc?

Kinda; since this is a list we'll make this `Computed` instead ([which we can do here](https://github.com/terraform-providers/terraform-provider-azurerm/pull/2123/files#diff-f1d68829a7ab03854b650243fd962709R225)) - given one item's being returned from Azure this means an item's being returned even if there's no items in that list; it also means we'll want to ensure we always set the items in that block (e.g. `upns`) - I'll push an updated review

**Comment by tombuildsstuff** [comment]
Ignoring a known test failure - the tests pass:

screenshot 2018-10-30 at 15 44 10

**Comment by tombuildsstuff** [comment]
hi @draggeta 

Just to let you know that this has been released as a part of v1.18 of the AzureRM Provider ([the full changelog is available here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/v1.18.0/CHANGELOG.md)). You can upgrade to this by specifying the version in the provider block (as shown below) and then running `terraform init -upgrade`

```
provider "azurerm" {
  version = "=1.18.0"
}
```

Thanks!

**Pull request merged**